content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1441 @@
+ipcMain.on('address-entered', async (event, address) =>  {
+  // ...
+})
+```
+
+In addition to the subscription to the ledger stream we also can subscribe the client to specific addresses, and we use this feature here to subscribe to an account address which we are going to prompt the user for:
+
+```javascript
+await client.request({
+  "command": "subscribe",
+  "streams": ["ledger"],
+  "accounts": [address]
+})
+```
+
+After this subscription our code attached listeners to the `ledgerClosed` and the `transactions` event. As soon as a `transaction` event is triggered, we do an `account_info` request to get the latest account status, as a transaction is an operation that changes the accounts state.
+
+In addition to the subscriptions we added each an initial `ledger` and `accountInfo` request to have some data at application startup, otherwise we would see empty fields until something happened on the ledger which would trigger one of our subscriptions.
+
+4. Now, add the following code to `preload.js`:
+
+```javascript
+const { contextBridge, ipcRenderer } = require('electron');
+
+contextBridge.exposeInMainWorld('electronAPI', {
+  onUpdateLedgerData: (callback) => {
+    ipcRenderer.on('update-ledger-data', callback)
+  },
+
+  // Step 3 code additions - start
+  onEnterAccountAddress: (address) => {
+    ipcRenderer.send('address-entered', address)
+  },
+  onUpdateAccountData: (callback) => {
+    ipcRenderer.on('update-account-data', callback)
+  }
+  //Step 3 code additions - end
+})
+```
+
+Here we can observe a notable difference to the previous step. Until now we just used `ipcRenderer` to pick up on events from the main logic, now we are using it bidirectional to send events from the frontend to the main logic:
+
+```javascript
+onEnterAccountAddress: (address) => {
+    ipcRenderer.send('address-entered', address)
+}
+```
+
+5. Then, modify the `view/template.html` by replacing the `<body>` with this markup: 
+
+```html
+<body>
+
+    <h3>Build a XRPL Wallet - Part 3/8</h3>
+
+    <fieldset>
+        <legend>Account info</legend>
+        Classic Address: <strong id="account-address-classic"></strong><br/>
+        X-Address: <strong id="account-address-x"></strong><br/>
+        XRP Balance: <strong id="account-balance"></strong><br/>
+        XRP Reserved: <strong id="account-reserve"></strong><br/>
+    </fieldset>
+
+    <fieldset>
+        <legend>Latest validated ledger stats</legend>
+        Ledger Index: <strong id="ledger-index"></strong><br/>
+        Ledger Hash: <strong id="ledger-hash"></strong><br/>
+        Close Time: <strong id="ledger-close-time"></strong><br/>
+    </fieldset>
+
+    <dialog id="account-address-dialog">
+        <form method="dialog">
+            <div>
+                <label for="address-input">Enter account address:</label>
+                <input type="text" id="address-input" name="address-input" />
+            </div>
+            <div>
+                <button type="reset">Reset</button>
```

Comment:
I'd recommend removing the reset button - it confused me at first because I clicked it by accident thinking it was confirm, and I don't think it's necessary since you can just delete what you typed.

Reply:
Done